sdks/java/core/src/main/java/org/apache/beam/sdk/io/GenerateSequence.java

```diff
@@ -0,0 +1,199 @@
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+/**
+ * A {@link PTransform} that produces longs starting from the given value, and either up to the
+ * given limit or until {@link Long#MAX_VALUE} / until the given time elapses.
+ *
+ * <p>The bounded {@link GenerateSequence} is implemented based on {@link OffsetBasedSource} and
+ * {@link OffsetBasedSource.OffsetBasedReader}, so it performs efficient initial splitting and it
+ * supports dynamic work rebalancing.
+ *
+ * <p>To produce a bounded {@code PCollection<Long>}:
+ *
+ * <pre>{@code
+ * Pipeline p = ...
+ * PCollection<Long> bounded = p.apply(GenerateSequence.from(0).to(1000));
+ * }</pre>
+ *
+ * <p>To produce an unbounded {@code PCollection<Long>}, simply do not specify {@link #to(long)},
+ * calling {@link #withTimestampFn(SerializableFunction)} to provide values with timestamps other
+ * than {@link Instant#now}.
+ *
+ * <pre>{@code
+ * Pipeline p = ...
+ *
+ * // To use processing time as the element timestamp.
+ * PCollection<Long> unbounded = p.apply(GenerateSequence.from(0));
+ * // Or, to use a provided function to set the element timestamp.
+ * PCollection<Long> unboundedWithTimestamps =
+ *     p.apply(GenerateSequence.from(0).withTimestampFn(someFn));
+ * }</pre>
+ *
+ * <p>In all cases, the sequence of numbers is generated in parallel, so there is no inherent
+ * ordering between the generated values - it is only guaranteed that all values in the given range
+ * will be present in the resulting {@link PCollection}.
+ */
+@AutoValue
+public abstract class GenerateSequence extends PTransform<PBegin, PCollection<Long>> {
+  abstract long getFrom();
+
+  abstract long getTo();
+
+  @Nullable
+  abstract SerializableFunction<Long, Instant> getTimestampFn();
+
+  abstract long getElementsPerPeriod();
+
+  @Nullable
+  abstract Duration getPeriod();
+
+  @Nullable
+  abstract Duration getMaxReadTime();
+
+  abstract Builder toBuilder();
+
+  @AutoValue.Builder
+  abstract static class Builder {
+    abstract Builder setFrom(long from);
+
+    abstract Builder setTo(long to);
+
+    abstract Builder setTimestampFn(SerializableFunction<Long, Instant> timestampFn);
+
+    abstract Builder setElementsPerPeriod(long elementsPerPeriod);
+
+    abstract Builder setPeriod(Duration period);
+
+    abstract Builder setMaxReadTime(Duration maxReadTime);
+
+    abstract GenerateSequence build();
+  }
+
+  /** Specifies the minimum number to generate (inclusive). */
+  public static GenerateSequence from(long from) {
+    checkArgument(from >= 0, "Value of from must be non-negative, but was: %s", from);
+    return new AutoValue_GenerateSequence.Builder()
+        .setFrom(from)
+        .setTo(-1)
+        .setElementsPerPeriod(0)
+        .build();
+  }
+
+  /** Specifies to generate the range [from, to). */
+  public static GenerateSequence fromTo(long from, long to) {
```

Comment:
`between`?

Then you get the more readable `Generate Sequence from x to y` and `Generate Sequence between x and y`